File: cloudflare/trafficpos.js
/**
 * Cloudflare Worker: Geo Locator (Preload/Cron Version)
 * 核心思路：通过后台 Cron 任务预加载数据到 KV，用户请求直接读取 KV。
 */

const DATA_SOURCES = {
  airports: "https://raw.githubusercontent.com/mwgg/Airports/master/airports.json",
  stations: "https://raw.githubusercontent.com/epcm/TrainVis/refs/heads/master/data_acquisition_and_processing/data/station_geo.json"
};

// KV 键名常量
const KEY_AIRPORTS = "airports_data";
const KEY_STATIONS = "stations_data";

export default {
  /**
   * 1. HTTP 请求处理 (面向用户)
   * 逻辑：优先读 KV -> 极速返回。只有 KV 为空时的紧急情况才去下载。
   */
  async fetch(request, env, _ctx) {
    if (!env.GEO_KV) return new Response("Error: KV Binding 'GEO_KV' missing", { status: 500 });

    const url = new URL(request.url);
    
    // 如果你是一个强迫症，想手动触发更新，可以保留这个秘密后门
    // 访问: https://你的域名/?action=force_update
    if (url.searchParams.get('action') === 'force_update') {
      await updateAllData(env.GEO_KV);
      return new Response("Data Force Updated Successfully!");
    }

    const lat = parseFloat(url.searchParams.get('lat'));
    const lon = parseFloat(url.searchParams.get('lon'));

    if (isNaN(lat) || isNaN(lon)) {
      return new Response(JSON.stringify({ error: "Provide 'lat' and 'lon'" }), { status: 400 });
    }

    // 从 KV 并行读取
    let [airportsData, stationsData] = await Promise.all([
      env.GEO_KV.get(KEY_AIRPORTS, { type: "json" }),
      env.GEO_KV.get(KEY_STATIONS, { type: "json" })
    ]);

    // 【兜底逻辑】万一 KV 是空的（刚部署完还没预加载），为了不报错，临时下载一次
    // 但正常情况下不会走到这里
    if (!airportsData || !stationsData) {
      console.log("KV empty, performing emergency fetch...");
      const freshData = await updateAllData(env.GEO_KV);
      airportsData = freshData.airports;
      stationsData = freshData.stations;
    }

    // 计算逻辑
    const nearestAirport = findNearestAirport(lat, lon, airportsData);
    const nearestStation = findNearestStation(lat, lon, stationsData);

    return new Response(JSON.stringify({
      input: { lat, lon },
      nearest_airport: nearestAirport,
      nearest_station: nearestStation
    }, null, 2), {
      headers: { 'content-type': 'application/json', 'Access-Control-Allow-Origin': '*' }
    });
  },

  /**
   * 2. Scheduled 定时任务 (面向后台维护)
   * 逻辑：定期（或手动）去 GitHub 下载数据并刷新 KV。
   * 这是实现“预加载”的关键。
   */
  async scheduled(_event, env, ctx) {
    ctx.waitUntil(updateAllData(env.GEO_KV));
  }
};

// --- 核心：数据更新函数 ---

async function updateAllData(kv) {
  console.log("Starting data update...");
  
  // 并行下载
  const [airportsResp, stationsResp] = await Promise.all([
    fetch(DATA_SOURCES.airports),
    fetch(DATA_SOURCES.stations)
  ]);

  if (!airportsResp.ok || !stationsResp.ok) {
    throw new Error("Failed to fetch data from GitHub");
  }

  const airportsData = await airportsResp.json();
  const stationsData = await stationsResp.json();

  // 写入 KV (不设置过期时间，或者设置很长，依靠 Cron 定期覆盖)
  // 这样数据永远存在，不会突然过期
  await Promise.all([
    kv.put(KEY_AIRPORTS, JSON.stringify(airportsData)),
    kv.put(KEY_STATIONS, JSON.stringify(stationsData))
  ]);

  console.log("Data updated successfully in KV.");
  
  return { airports: airportsData, stations: stationsData };
}

// --- 计算工具函数 (保持不变) ---
function findNearestAirport(targetLat, targetLon, data) {
  let minDist = Infinity; let nearest = null;
  for (const icao in data) {
    const d = data[icao];
    if (!d.lat || !d.lon || !d.iata) continue;
    const dist = getDistanceFromLatLonInKm(targetLat, targetLon, d.lat, d.lon);
    if (dist < minDist) { minDist = dist; nearest = { code: d.iata, name: d.name, dist: parseFloat(dist.toFixed(2)) }; }
  }
  return nearest;
}
function findNearestStation(targetLat, targetLon, data) {
  let minDist = Infinity; let nearest = null;
  for (const [name, coords] of Object.entries(data)) {
    if (!Array.isArray(coords) || coords.length < 2) continue;
    const dist = getDistanceFromLatLonInKm(targetLat, targetLon, coords[1], coords[0]);
    if (dist < minDist) { minDist = dist; nearest = { name: name, dist: parseFloat(dist.toFixed(2)) }; }
  }
  return nearest;
}
function getDistanceFromLatLonInKm(lat1, lon1, lat2, lon2) {
  const R = 6371; const dLat = deg2rad(lat2 - lat1); const dLon = deg2rad(lon2 - lon1);
  const a = Math.sin(dLat / 2) * Math.sin(dLat / 2) + Math.cos(deg2rad(lat1)) * Math.cos(deg2rad(lat2)) * Math.sin(dLon / 2) * Math.sin(dLon / 2);
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}
function deg2rad(deg) { return deg * (Math.PI / 180); }

File: backend/internal/handler/trafficpos.go
package handler

import (
	"encoding/json"
	"fmt"
	"math"
	"net/http"
	"os"
	"path/filepath"

	"github.com/gin-gonic/gin"
)

// TrafficPosResponse 响应结构
type TrafficPosResponse struct {
	Input          LatLonInput   `json:"input"`
	NearestAirport *NearestPoint `json:"nearest_airport,omitempty"`
	NearestStation *NearestPoint `json:"nearest_station,omitempty"`
}

type LatLonInput struct {
	Lat float64 `json:"lat"`
	Lon float64 `json:"lon"`
}

type NearestPoint struct {
	Code   string  `json:"code,omitempty"`
	Name   string  `json:"name"`
	DistKm float64 `json:"dist_km"`
}

// TrafficPosData 存储机场和站点数据
type TrafficPosData struct {
	Airports map[string][2]float64 `json:"airports"` // IATA: [lat, lon]
	Stations map[string][2]float64 `json:"stations"` // 站名: [lon, lat]
}

var trafficPosData *TrafficPosData

// LoadTrafficPosData 加载机场和高铁站数据
func LoadTrafficPosData(configPath string) error {
	trafficPosData = &TrafficPosData{
		Airports: make(map[string][2]float64),
		Stations: make(map[string][2]float64),
	}

	// 加载机场数据
	airportsPath := filepath.Join(configPath, "airports.json")
	if err := loadJSONFile(airportsPath, &trafficPosData.Airports); err != nil {
		return err
	}

	// 加载高铁站数据
	stationsPath := filepath.Join(configPath, "station_geo.json")
	if err := loadJSONFile(stationsPath, &trafficPosData.Stations); err != nil {
		return err
	}

	return nil
}

// loadJSONFile 加载JSON文件
func loadJSONFile(path string, v interface{}) error {
	data, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	return json.Unmarshal(data, v)
}

// GetTrafficPos 获取最近的机场和高铁站
func GetTrafficPos(c *gin.Context) {
	latStr := c.Query("lat")
	lonStr := c.Query("lon")

	if latStr == "" || lonStr == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "请提供有效的 'lat' 和 'lon' 参数",
		})
		return
	}

	var lat, lon float64
	if _, err := fmt.Sscanf(latStr, "%f", &lat); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "无效的纬度参数",
		})
		return
	}
	if _, err := fmt.Sscanf(lonStr, "%f", &lon); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "无效的经度参数",
		})
		return
	}

	if trafficPosData == nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "数据未初始化",
		})
		return
	}

	response := TrafficPosResponse{
		Input: LatLonInput{Lat: lat, Lon: lon},
	}

	// 查找最近机场
	if nearestAirport := findNearestAirport(lat, lon); nearestAirport != nil {
		response.NearestAirport = nearestAirport
	}

	// 查找最近高铁站
	if nearestStation := findNearestStation(lat, lon); nearestStation != nil {
		response.NearestStation = nearestStation
	}

	c.JSON(http.StatusOK, response)
}

// findNearestAirport 查找最近机场
func findNearestAirport(targetLat, targetLon float64) *NearestPoint {
	if trafficPosData == nil || len(trafficPosData.Airports) == 0 {
		return nil
	}

	var minDist float64 = math.MaxFloat64
	var nearest *NearestPoint

	for iata, coords := range trafficPosData.Airports {
		if len(coords) < 2 {
			continue
		}

		dist := getDistanceFromLatLonInKm(targetLat, targetLon, coords[0], coords[1])
		if dist < minDist {
			minDist = dist
			nearest = &NearestPoint{
				Code:   iata,
				Name:   iata,                       // 使用IATA代码作为名称，因为没有其他信息
				DistKm: math.Round(dist*100) / 100, // 保留2位小数
			}
		}
	}

	return nearest
}

// findNearestStation 查找最近高铁站
func findNearestStation(targetLat, targetLon float64) *NearestPoint {
	if trafficPosData == nil || len(trafficPosData.Stations) == 0 {
		return nil
	}

	var minDist float64 = math.MaxFloat64
	var nearest *NearestPoint

	for name, coords := range trafficPosData.Stations {
		if len(coords) < 2 {
			continue
		}

		// 注意：高铁站数据是[lon, lat]格式
		dist := getDistanceFromLatLonInKm(targetLat, targetLon, coords[1], coords[0])
		if dist < minDist {
			minDist = dist
			nearest = &NearestPoint{
				Name:   name,
				DistKm: math.Round(dist*100) / 100, // 保留2位小数
			}
		}
	}

	return nearest
}

// getDistanceFromLatLonInKm Haversine公式计算距离
func getDistanceFromLatLonInKm(lat1, lon1, lat2, lon2 float64) float64 {
	const R = 6371 // 地球半径（千米）

	dLat := deg2rad(lat2 - lat1)
	dLon := deg2rad(lon2 - lon1)

	a := math.Sin(dLat/2)*math.Sin(dLat/2) +
		math.Cos(deg2rad(lat1))*math.Cos(deg2rad(lat2))*
			math.Sin(dLon/2)*math.Sin(dLon/2)

	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
	return R * c
}

// deg2rad 角度转弧度
func deg2rad(deg float64) float64 {
	return deg * (math.Pi / 180)
}

